src/Storages/StorageS3.cpp

```diff
@@ -322,6 +339,161 @@ class StorageS3BlockOutputStream : public IBlockOutputStream
+        {
+            getWriterForPartition(partition_id)->write(sub_blocks[index]);
+        }
+    }
+
+    void writePrefix() override
+    {
+        /// `writePrefix()` for each child writer is called in `getWriterForPartition()`
+        /// on writer creation.
+    }
+
+    void flush() override
+    {
+        for (auto & [partition_id, writer] : writers)
+        {
+            writer->flush();
+        }
+    }
+
+    void writeSuffix() override
+    {
+        for (auto & [partition_id, writer] : writers)
+        {
+            writer->writeSuffix();
+        }
+    }
+
+private:
+    const String format;
+    const Block sample_block;
+    ContextPtr context;
+    const CompressionMethod compression_method;
+    std::shared_ptr<Aws::S3::S3Client> client;
+    const String bucket;
+    const String key;
+    size_t min_upload_part_size;
+    size_t max_single_part_upload_size;
+
+    ExpressionActionsPtr partition_by_expr;
+    String partition_by_column_name;
+
+    std::unordered_map<String, BlockOutputStreamPtr> writers;
+
+    static String replaceWildcards(const String & input, const String & partition_id)
+    {
+        return boost::replace_all_copy(input, PARTITION_ID_WILDCARD, partition_id);
+    }
+
+    BlockOutputStreamPtr getWriterForPartition(const String & partition_id)
+    {
+        if (writers.count(partition_id) == 0)
+        {
+            writers.emplace(partition_id, std::make_shared<StorageS3BlockOutputStream>(
+                format,
+                sample_block,
+                context,
+                compression_method,
+                client,
+                replaceWildcards(bucket, partition_id),
+                replaceWildcards(key, partition_id),
+                min_upload_part_size,
+                max_single_part_upload_size
+            )).first->second->writePrefix();
+        }
+
+        return writers[partition_id];
+    }
+
+    static void validateParitionKey(const StringRef & str)
+    {
+        if (str.size == 0)
+            throw DB::Exception(ErrorCodes::CANNOT_PARSE_TEXT, "Illegal empty partition key");
+
+        static constexpr std::array<char, 8> valid_chars = {'/', '.', '!', '(', ')', '\'', '_', '-'};
+        for (const char * p = str.data; static_cast<size_t>(p - str.data) < str.size; ++p)
+        {
+            bool is_valid = isWordCharASCII(*p) || std::any_of(valid_chars.begin(), valid_chars.end(), [p](char v){ return v == *p; });
+            if (unlikely(!is_valid))
+            {
+                /// need to convert to uint32 because uint8 can't be passed to format due to "mixing character types is disallowed"
+                UInt32 invalid_char_byte = static_cast<UInt32>(static_cast<UInt8>(*p));
+                throw DB::Exception(
+                    ErrorCodes::CANNOT_PARSE_TEXT,
+                    "Illegal character '\\x{0:02x}' in partition key. Allowed symbols: [a-zA-Z0-9{1}]", invalid_char_byte, fmt::join(valid_chars, ""));
```

Comment:
I guess, that this is not very good, alphabetic characters from other languages are disabled now.

Reply:
I would also add links `https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html` and `https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html` to the comments in this case.